File: macjet/ui/detail_panel.py
"""
MacJet — Detail Panel (Inspector Rail)
Shows sparklines, why-hot analysis, process metadata, and action shortcuts
for the selected process or group.
"""
from __future__ import annotations

from textual.widget import Widget
from textual.app import ComposeResult
from textual.widgets import Static

from ..collectors.process_collector import ProcessGroup, ProcessInfo
from ..collectors.metrics_history import MetricsHistory, ReclaimCandidate

# Optional imports for browser/IDE context
try:
    from ..inspectors.browser_inspector import BrowserContext
except ImportError:
    BrowserContext = None

try:
    from ..inspectors.ide_inspector import IDEContext
except ImportError:
    IDEContext = None

try:
    from ..inspectors.chrome_tab_mapper import ChromeSnapshot
except ImportError:
    ChromeSnapshot = None


def _format_mem(mb: float) -> str:
    if mb >= 1024:
        return f"{mb / 1024:.1f}GB"
    return f"{mb:.0f}MB"


def _format_duration(seconds: float) -> str:
    if seconds < 60:
        return f"{int(seconds)}s"
    elif seconds < 3600:
        return f"{int(seconds / 60)}m"
    else:
        h = int(seconds / 3600)
        m = int((seconds % 3600) / 60)
        return f"{h}h{m}m" if m else f"{h}h"


class DetailPanel(Widget):
    """Right-side inspector panel showing deep context for selected process/group."""

    DEFAULT_CSS = """
    DetailPanel {
        width: 38;
        border-left: solid #1A2540;
        background: #10182B;
        padding: 0 1;
    }
    DetailPanel #detail-title {
        height: 1;
        text-style: bold;
        color: #60A5FA;
    }
    DetailPanel #detail-body {
        height: 1fr;
        color: #E6ECFF;
    }
    """

    def compose(self) -> ComposeResult:
        yield Static("  [#60A5FA]Inspector[/]", id="detail-title")
        yield Static("  [#7F8DB3]Select a process to inspect[/]", id="detail-body")

    def _update(self, title: str, body: str):
        try:
            self.query_one("#detail-title", Static).update(title)
            self.query_one("#detail-body", Static).update(body)
        except Exception:
            pass

    def show_empty(self):
        self._update(
            "  [#60A5FA]Inspector[/]",
            "  [#7F8DB3]Select a process to inspect[/]"
        )

    def show_message(self, title: str, message: str):
        """Show a simple message in the inspector."""
        self._update(f"  [#FF8A4C]{title}[/]", f"  [#7F8DB3]{message}[/]")

    def show_process(self, proc: ProcessInfo, group: ProcessGroup | None = None,
                     metrics: MetricsHistory | None = None):
        """Show details for a single process."""
        title = f"  📋 {proc.context_label or proc.name}"

        lines = []
        # Sparkline
        if metrics:
            spark = metrics.sparkline(proc.pid, width=28)
            lines.append(f"  [#45D6FF]{spark}[/]")
            lines.append(f"  [#7F8DB3]60s CPU trend[/]")
            lines.append("")

        lines.append(f"  [#7F8DB3]PID:[/]     [#E6ECFF]{proc.pid}[/]")
        lines.append(f"  [#7F8DB3]CPU:[/]     [#E6ECFF]{proc.cpu_percent:.1f}%[/]")
        lines.append(f"  [#7F8DB3]Memory:[/]  [#E6ECFF]{_format_mem(proc.memory_mb)}[/]")
        lines.append(f"  [#7F8DB3]Threads:[/] [#E6ECFF]{proc.num_threads}[/]")
        lines.append(f"  [#7F8DB3]Status:[/]  [#E6ECFF]{proc.status}[/]")

        if proc.launch_age_s > 0:
            lines.append(f"  [#7F8DB3]Age:[/]     [#E6ECFF]{_format_duration(proc.launch_age_s)}[/]")

        if proc.exe:
            exe_short = proc.exe if len(proc.exe) <= 30 else "…" + proc.exe[-29:]
            lines.append(f"  [#7F8DB3]Exe:[/]  [#E6ECFF]{exe_short}[/]")

        if proc.role_type:
            lines.append(f"  [#7F8DB3]Role:[/]    [#A78BFA]{proc.role_type}[/]")

        if proc.is_system:
            lines.append(f"  [#7F8DB3]Type:[/]    [#FF8A4C]System[/]")

        if proc.energy_impact:
            energy_color = "#FF4D6D" if proc.energy_impact == "HIGH" else (
                "#FF8A4C" if proc.energy_impact == "MED" else "#32D583"
            )
            lines.append(f"  [#7F8DB3]Energy:[/]  [{energy_color}]{proc.energy_impact}[/]")

        # Memory trend
        if metrics:
            growth = metrics.memory_growth_rate(proc.pid)
            if abs(growth) > 1:
                growth_color = "#FF4D6D" if growth > 10 else "#FDBA35"
                lines.append(f"  [#7F8DB3]Δ Mem:[/]   [{growth_color}]{growth:+.0f}MB/min[/]")

        lines.append("")
        lines.append("  [#7F8DB3]─── Actions ───────────[/]")
        lines.append("  [#60A5FA]k[/] Kill  [#60A5FA]K[/] Force Kill")
        lines.append("  [#60A5FA]p[/] Profile  [#60A5FA]z[/] Suspend")

        self._update(title, "\n".join(lines))

    def show_group(self, group: ProcessGroup, metrics: MetricsHistory | None = None):
        """Show details for a process group with sparkline and role breakdown."""
        title = f"  📋 {group.name} ({len(group.processes)})"

        lines = []
        # Group sparkline
        if metrics:
            pids = [p.pid for p in group.processes]
            spark = metrics.sparkline_for_group(pids, width=28)
            lines.append(f"  [#45D6FF]{spark}[/]")
            lines.append(f"  [#7F8DB3]60s CPU trend (group)[/]")
            lines.append("")

        lines.append(f"  [#7F8DB3]CPU:[/]     [#E6ECFF]{group.total_cpu:.1f}%[/]")
        lines.append(f"  [#7F8DB3]Memory:[/]  [#E6ECFF]{_format_mem(group.total_memory_mb)}[/]")
        lines.append(f"  [#7F8DB3]Procs:[/]   [#E6ECFF]{len(group.processes)}[/]")

        # Memory trend
        if metrics:
            total_growth = sum(
                max(0, metrics.memory_growth_rate(p.pid))
                for p in group.processes
            )
            if total_growth > 1:
                growth_color = "#FF4D6D" if total_growth > 50 else "#FDBA35"
                lines.append(f"  [#7F8DB3]Δ Mem:[/]   [{growth_color}]{total_growth:+.0f}MB/min[/]")

        # Why hot analysis
        if group.total_cpu > 10:
            lines.append("")
            lines.append("  [#FF8A4C]🔥 Why hot:[/]")
            reasons = self._analyze_why_hot(group, metrics)
            for reason in reasons:
                lines.append(f"  [#E6ECFF]{reason}[/]")

        # Role breakdown
        roles: dict[str, tuple[int, float, float]] = {}
        for p in group.processes:
            role = p.role_type or "main"
            if role not in roles:
                roles[role] = (0, 0.0, 0.0)
            count, cpu, mem = roles[role]
            roles[role] = (count + 1, cpu + p.cpu_percent, mem + p.memory_mb)

        if len(roles) > 1:
            lines.append("")
            lines.append("  [#7F8DB3]─── Breakdown ─────────[/]")
            sorted_roles = sorted(roles.items(), key=lambda x: x[1][1], reverse=True)
            for role, (count, cpu, mem) in sorted_roles:
                label = role.capitalize()
                lines.append(f"  [#E6ECFF]{label} ×{count}[/]  [#7F8DB3]{cpu:.1f}%  {_format_mem(mem)}[/]")

        lines.append("")
        lines.append("  [#7F8DB3]─── Actions ───────────[/]")
        lines.append("  [#60A5FA]k[/] Kill  [#60A5FA]K[/] Force Kill")
        lines.append("  [#60A5FA]p[/] Profile  [#60A5FA]z[/] Suspend")

        self._update(title, "\n".join(lines))

    def _analyze_why_hot(self, group: ProcessGroup, metrics: MetricsHistory | None) -> list[str]:
        """Generate why-hot analysis reasons."""
        reasons = []

        if group.total_cpu > 80:
            reasons.append("Sustained high CPU usage")
        elif group.total_cpu > 30:
            reasons.append("Elevated CPU usage")

        # Check for process storms
        renderers = sum(1 for p in group.processes if p.role_type == "renderer")
        if renderers > 10:
            reasons.append(f"Renderer storm: {renderers} renderers")

        # Memory growth
        if metrics:
            growth = sum(
                max(0, metrics.memory_growth_rate(p.pid))
                for p in group.processes
            )
            if growth > 20:
                reasons.append(f"Memory growing +{growth:.0f}MB/min")

        # Hidden/background
        if all(not hasattr(p, 'is_hidden') or p.is_hidden for p in group.processes):
            reasons.append("Running in background")

        # High energy
        high_energy = sum(1 for p in group.processes if p.energy_impact == "HIGH")
        if high_energy:
            reasons.append(f"{high_energy} high-energy processes")

        if not reasons:
            reasons.append("Active usage")

        return reasons

    def show_reclaim(self, candidate: ReclaimCandidate, group: ProcessGroup,
                     metrics: MetricsHistory | None = None):
        """Show Reclaim details for a kill candidate."""
        risk_colors = {
            "safe": "#32D583",
            "review": "#FDBA35",
            "danger": "#FF4D6D",
        }
        risk_color = risk_colors.get(candidate.risk, "#7F8DB3")

        title = f"  ⚡ {candidate.app_name}"
        lines = []

        # Sparkline
        if metrics:
            pids = [p.pid for p in group.processes]
            spark = metrics.sparkline_for_group(pids, width=28)
            lines.append(f"  [#45D6FF]{spark}[/]")
            lines.append("")

        lines.append(f"  [#7F8DB3]Score:[/]   [#E6ECFF]{candidate.score}/100[/]")
        lines.append(f"  [#7F8DB3]Risk:[/]    [{risk_color}]{candidate.risk.upper()}[/]")
        lines.append(f"  [#7F8DB3]Reclaim:[/] [#E6ECFF]~{candidate.reclaim_cpu:.0f}% CPU / {_format_mem(candidate.reclaim_mem_mb)}[/]")
        lines.append("")
        lines.append(f"  [#FF8A4C]Reason:[/]")
        lines.append(f"  [#E6ECFF]{candidate.reason}[/]")
        lines.append("")
        lines.append(f"  [#7F8DB3]Suggested:[/] [#60A5FA]{candidate.suggested_action}[/]")
        lines.append(f"  [#7F8DB3]Children:[/]  [#E6ECFF]{candidate.child_count}[/]")

        if candidate.launch_age_s > 0:
            lines.append(f"  [#7F8DB3]Age:[/]       [#E6ECFF]{_format_duration(candidate.launch_age_s)}[/]")

        lines.append("")
        lines.append("  [#7F8DB3]─── Actions ───────────[/]")
        lines.append("  [#60A5FA]k[/] Terminate  [#60A5FA]K[/] Force Kill")
        lines.append("  [#60A5FA]p[/] Sample     [#60A5FA]z[/] Suspend")

        self._update(title, "\n".join(lines))

    def show_browser(self, group: ProcessGroup, browser_ctx):
        """Show browser-specific details."""
        title = f"  🌐 {browser_ctx.app_name}"
        lines = []
        lines.append(f"  [#7F8DB3]CPU:[/]     [#E6ECFF]{group.total_cpu:.1f}%[/]")
        lines.append(f"  [#7F8DB3]Memory:[/]  [#E6ECFF]{_format_mem(group.total_memory_mb)}[/]")
        lines.append(f"  [#7F8DB3]Windows:[/] [#E6ECFF]{browser_ctx.window_count}[/]")
        lines.append(f"  [#7F8DB3]Tabs:[/]    [#E6ECFF]{browser_ctx.tab_count}[/]")
        lines.append("")

        if browser_ctx.active_tab:
            lines.append("  [#60A5FA]▶ Active Tab:[/]")
            at = browser_ctx.active_tab
            lines.append(f"    [#E6ECFF]{at.title[:30]}[/]")
            lines.append("")

        lines.append("  [#7F8DB3]All Tabs:[/]")
        for i, tab in enumerate(browser_ctx.tabs[:10]):
            marker = "[#60A5FA]→[/] " if tab.is_active else "  "
            tab_title = tab.title[:28] if tab.title else "(untitled)"
            lines.append(f"  {marker}{i+1}. [#E6ECFF]{tab_title}[/]")

        remaining = browser_ctx.tab_count - 10
        if remaining > 0:
            lines.append(f"    [#7F8DB3]...and {remaining} more[/]")

        self._update(title, "\n".join(lines))

    def show_ide(self, proc: ProcessInfo, ide_ctx):
        """Show IDE-specific details."""
        title = f"  💻 {ide_ctx.ide_name}"
        lines = []
        if ide_ctx.project_name:
            lines.append(f"  [#7F8DB3]Project:[/] [#E6ECFF]{ide_ctx.project_name}[/]")
        if ide_ctx.project_path:
            lines.append(f"  [#7F8DB3]Path:[/]    [#E6ECFF]{ide_ctx.project_path}[/]")
        if ide_ctx.active_file:
            lines.append(f"  [#7F8DB3]File:[/]    [#E6ECFF]{ide_ctx.active_file}[/]")
        if ide_ctx.window_title:
            lines.append(f"  [#7F8DB3]Window:[/]  [#E6ECFF]{ide_ctx.window_title}[/]")
        lines.append("")
        lines.append(f"  [#7F8DB3]CPU:[/]     [#E6ECFF]{proc.cpu_percent:.1f}%[/]")
        lines.append(f"  [#7F8DB3]Memory:[/]  [#E6ECFF]{_format_mem(proc.memory_mb)}[/]")
        lines.append(f"  [#7F8DB3]PID:[/]     [#E6ECFF]{proc.pid}[/]")

        self._update(title, "\n".join(lines))

    def show_chrome_cdp(self, group: ProcessGroup, snapshot):
        """Show Chrome tabs with JS heap from CDP."""
        title = f"  🌐 Chrome ({snapshot.total_tabs} tabs)"
        lines = []
        lines.append(f"  [#7F8DB3]CPU:[/]     [#E6ECFF]{group.total_cpu:.1f}%[/]")
        lines.append(f"  [#7F8DB3]Memory:[/]  [#E6ECFF]{_format_mem(group.total_memory_mb)}[/]")
        lines.append(f"  [#7F8DB3]JS Heap:[/] [#E6ECFF]{snapshot.total_js_heap_mb:.0f}MB[/]")
        lines.append("")
        lines.append("  [#7F8DB3]Tabs (by JS heap):[/]")

        for tab in snapshot.tabs[:12]:
            heap_str = f"{tab.js_heap_mb:.0f}MB" if tab.js_heap_mb > 0.5 else "<1MB"
            tab_title = tab.title[:26] if tab.title else "(untitled)"

            if tab.js_heap_mb > 200:
                icon = "[#FF4D6D]●[/]"
            elif tab.js_heap_mb > 50:
                icon = "[#FF8A4C]●[/]"
            elif tab.js_heap_mb > 10:
                icon = "[#FDBA35]●[/]"
            else:
                icon = "[#32D583]●[/]"

            lines.append(f"  {icon} [#E6ECFF]{tab_title}[/]")
            lines.append(f"     [#7F8DB3]{heap_str}[/]")

        remaining = snapshot.total_tabs - 12
        if remaining > 0:
            lines.append(f"  [#7F8DB3]...and {remaining} more[/]")

        self._update(title, "\n".join(lines))

File: src/ui/detail_panel.rs
/// MacJet — Detail Panel (Inspector)
///
/// Shows dynamic metadata, sparklines, "Why Hot" analysis, role breakdowns,
/// memory growth rates, and action shortcuts for the selected process/group.
///
/// This panel occupies the right 38-char column of the main layout.
use crate::collectors::metrics_history::{MetricsHistory, ReclaimCandidate};
use crate::collectors::process_collector::{ProcessGroup, ProcessInfo};
use crate::ui::styles;
use ratatui::{
    buffer::Buffer,
    layout::Rect,
    style::{Color, Modifier, Style},
    text::{Line, Span},
    widgets::{Block, Borders, Paragraph, Widget},
};

pub struct DetailPanelWidget<'a> {
    pub process: Option<&'a ProcessInfo>,
    pub group: Option<&'a ProcessGroup>,
    pub reclaim_candidate: Option<&'a ReclaimCandidate>,
    pub reclaim_group: Option<&'a ProcessGroup>,
    pub metrics: &'a MetricsHistory,
}

impl<'a> DetailPanelWidget<'a> {
    pub fn new(
        process: Option<&'a ProcessInfo>,
        group: Option<&'a ProcessGroup>,
        metrics: &'a MetricsHistory,
    ) -> Self {
        Self {
            process,
            group,
            reclaim_candidate: None,
            reclaim_group: None,
            metrics,
        }
    }

    pub fn from_reclaim(
        candidate: Option<&'a ReclaimCandidate>,
        group: Option<&'a ProcessGroup>,
        metrics: &'a MetricsHistory,
    ) -> Self {
        Self {
            process: None,
            group: None,
            reclaim_candidate: candidate,
            reclaim_group: group,
            metrics,
        }
    }

    fn format_mem(mb: f64) -> String {
        if mb >= 1024.0 {
            format!("{:.1}GB", mb / 1024.0)
        } else {
            format!("{:.0}MB", mb)
        }
    }

    fn format_duration(seconds: f64) -> String {
        if seconds < 60.0 {
            format!("{:.0}s", seconds)
        } else if seconds < 3600.0 {
            format!("{:.0}m", seconds / 60.0)
        } else {
            let h = (seconds / 3600.0) as u32;
            let m = ((seconds % 3600.0) / 60.0) as u32;
            if m > 0 {
                format!("{}h{}m", h, m)
            } else {
                format!("{}h", h)
            }
        }
    }

    /// Analyzes why a hot group is burning CPU and returns human-readable reasons.
    fn why_hot(group: &ProcessGroup, metrics: &MetricsHistory) -> Vec<String> {
        let mut reasons = Vec::new();

        if group.total_cpu > 80.0 {
            reasons.push("Sustained high CPU usage".to_string());
        } else if group.total_cpu > 30.0 {
            reasons.push("Elevated CPU usage".to_string());
        }

        let renderers = group
            .processes
            .iter()
            .filter(|p| p.role_type.as_str() == "renderer")
            .count();
        if renderers > 10 {
            reasons.push(format!("Renderer storm: {} renderers", renderers));
        }

        let growth: f64 = group
            .processes
            .iter()
            .map(|p| metrics.memory_growth_rate(p.pid, 60.0).max(0.0))
            .sum();
        if growth > 20.0 {
            reasons.push(format!("Memory growing +{:.0}MB/min", growth));
        }

        if group.processes.iter().all(|p| p.is_hidden || p.is_system) {
            reasons.push("Running in background".to_string());
        }

        let high_energy = group
            .processes
            .iter()
            .filter(|p| p.energy_impact.as_str() == "HIGH")
            .count();
        if high_energy > 0 {
            reasons.push(format!("{} high-energy processes", high_energy));
        }

        if reasons.is_empty() {
            reasons.push("Active usage".to_string());
        }

        reasons
    }

    fn sparkline_for_pid(&self, pid: u32, width: usize) -> String {
        self.metrics.sparkline(pid, width, "cpu")
    }

    fn sparkline_for_group(&self, pids: &[u32], width: usize) -> String {
        self.metrics.sparkline_for_group(pids, width)
    }
}

impl<'a> Widget for DetailPanelWidget<'a> {
    fn render(self, area: Rect, buf: &mut Buffer) {
        let block = Block::default()
            .borders(Borders::LEFT)
            .border_style(Style::default().fg(styles::BORDER_DIM))
            .style(Style::default().bg(styles::BG_HEADER));

        let title = if let Some(c) = self.reclaim_candidate {
            format!("  ⚡ {}", c.app_name)
        } else if let Some(g) = self.group {
            format!("  📋 {} ({})", g.name, g.processes.len())
        } else if let Some(p) = self.process {
            let name = if p.context_label.is_empty() {
                p.name.as_str()
            } else {
                p.context_label.as_str()
            };
            format!("  📋 {}", name)
        } else {
            "  Inspector".to_string()
        };

        let mut lines: Vec<Line> = Vec::new();

        // --- Title Line ---
        lines.push(Line::from(Span::styled(
            title,
            Style::default()
                .fg(styles::ACCENT_BLUE)
                .add_modifier(Modifier::BOLD),
        )));

        // === Reclaim Candidate Mode ===
        if let Some(c) = self.reclaim_candidate {
            if let Some(g) = self.reclaim_group {
                let pids: Vec<u32> = g.processes.iter().map(|p| p.pid).collect();
                lines.push(Line::from(vec![
                    Span::raw("  "),
                    Span::styled(
                        self.sparkline_for_group(&pids, 28),
                        Style::default().fg(styles::ACCENT_CYAN),
                    ),
                ]));
                lines.push(Line::from(""));
            }

            let risk_color = match c.risk.as_str() {
                "safe" => styles::ACCENT_GREEN,
                "review" => styles::ACCENT_AMBER,
                "danger" => styles::ACCENT_RED,
                _ => styles::TEXT_DIM,
            };

            kv(
                &mut lines,
                "Score:",
                &format!("{}/100", c.score),
                styles::TEXT_BRIGHT,
            );
            kv(&mut lines, "Risk:", &c.risk.to_uppercase(), risk_color);
            kv(
                &mut lines,
                "Reclaim:",
                &format!(
                    "~{:.0}% CPU / {}",
                    c.reclaim_cpu,
                    Self::format_mem(c.reclaim_mem_mb)
                ),
                styles::TEXT_BRIGHT,
            );

            lines.push(Line::from(""));
            lines.push(Line::from(vec![
                Span::raw("  "),
                Span::styled("Reason:", Style::default().fg(risk_color)),
            ]));
            lines.push(Line::from(vec![
                Span::raw("  "),
                Span::styled(
                    c.reason.to_string(),
                    Style::default().fg(styles::TEXT_BRIGHT),
                ),
            ]));

            lines.push(Line::from(""));
            kv(
                &mut lines,
                "Suggested:",
                c.suggested_action.as_str(),
                styles::ACCENT_BLUE,
            );
            kv(
                &mut lines,
                "Children:",
                &c.child_count.to_string(),
                styles::TEXT_BRIGHT,
            );

            if c.launch_age_s > 0.0 {
                kv(
                    &mut lines,
                    "Age:",
                    &Self::format_duration(c.launch_age_s),
                    styles::TEXT_BRIGHT,
                );
            }
        }
        // === Group Mode ===
        else if let Some(g) = self.group {
            let pids: Vec<u32> = g.processes.iter().map(|p| p.pid).collect();
            lines.push(Line::from(vec![
                Span::raw("  "),
                Span::styled(
                    self.sparkline_for_group(&pids, 28),
                    Style::default().fg(styles::ACCENT_CYAN),
                ),
            ]));
            lines.push(Line::from(vec![
                Span::raw("  "),
                Span::styled(
                    "60s CPU trend (group)",
                    Style::default().fg(styles::TEXT_DIM),
                ),
            ]));
            lines.push(Line::from(""));

            kv(
                &mut lines,
                "CPU:",
                &format!("{:.1}%", g.total_cpu),
                styles::TEXT_BRIGHT,
            );
            kv(
                &mut lines,
                "Memory:",
                &Self::format_mem(g.total_memory_mb),
                styles::TEXT_BRIGHT,
            );
            kv(
                &mut lines,
                "Procs:",
                &g.processes.len().to_string(),
                styles::TEXT_BRIGHT,
            );

            // Memory growth rate
            let total_growth: f64 = g
                .processes
                .iter()
                .map(|p| self.metrics.memory_growth_rate(p.pid, 60.0).max(0.0))
                .sum();
            if total_growth > 1.0 {
                let color = if total_growth > 50.0 {
                    styles::ACCENT_RED
                } else {
                    styles::ACCENT_AMBER
                };
                kv(
                    &mut lines,
                    "Δ Mem:",
                    &format!("+{:.0}MB/min", total_growth),
                    color,
                );
            }

            // Why Hot analysis
            if g.total_cpu > 10.0 {
                lines.push(Line::from(""));
                lines.push(Line::from(vec![
                    Span::raw("  "),
                    Span::styled("🔥 Why hot:", Style::default().fg(Color::Rgb(255, 138, 76))),
                ]));
                for reason in Self::why_hot(g, self.metrics) {
                    lines.push(Line::from(vec![
                        Span::raw("  "),
                        Span::styled(reason, Style::default().fg(styles::TEXT_BRIGHT)),
                    ]));
                }
            }

            // Role breakdown
            let mut roles: std::collections::BTreeMap<String, (usize, f64, f64)> =
                std::collections::BTreeMap::new();
            for p in &g.processes {
                let role = if p.role_type.is_empty() {
                    "main".to_string()
                } else {
                    p.role_type.to_string()
                };
                let entry = roles.entry(role).or_insert((0, 0.0, 0.0));
                entry.0 += 1;
                entry.1 += p.cpu_percent;
                entry.2 += p.memory_mb;
            }

            if roles.len() > 1 {
                lines.push(Line::from(""));
                section_header(&mut lines, "─── Breakdown ─────────");
                let mut sorted: Vec<_> = roles.into_iter().collect();
                sorted.sort_by(|a, b| {
                    b.1 .1
                        .partial_cmp(&a.1 .1)
                        .unwrap_or(std::cmp::Ordering::Equal)
                });
                for (role, (count, cpu, mem)) in sorted {
                    lines.push(Line::from(vec![
                        Span::raw("  "),
                        Span::styled(
                            format!("{} ×{}  ", titlecase(&role), count),
                            Style::default().fg(styles::TEXT_BRIGHT),
                        ),
                        Span::styled(
                            format!("{:.1}%  {}", cpu, Self::format_mem(mem)),
                            Style::default().fg(styles::TEXT_DIM),
                        ),
                    ]));
                }
            }
        }
        // === Single Process Mode ===
        else if let Some(p) = self.process {
            lines.push(Line::from(vec![
                Span::raw("  "),
                Span::styled(
                    self.sparkline_for_pid(p.pid, 28),
                    Style::default().fg(styles::ACCENT_CYAN),
                ),
            ]));
            lines.push(Line::from(vec![
                Span::raw("  "),
                Span::styled("60s CPU trend", Style::default().fg(styles::TEXT_DIM)),
            ]));
            lines.push(Line::from(""));

            kv(&mut lines, "PID:", &p.pid.to_string(), styles::TEXT_BRIGHT);
            kv(
                &mut lines,
                "CPU:",
                &format!("{:.1}%", p.cpu_percent),
                styles::TEXT_BRIGHT,
            );
            kv(
                &mut lines,
                "Memory:",
                &Self::format_mem(p.memory_mb),
                styles::TEXT_BRIGHT,
            );
            kv(
                &mut lines,
                "Threads:",
                &p.num_threads.to_string(),
                styles::TEXT_BRIGHT,
            );
            kv(
                &mut lines,
                "Status:",
                p.status.as_str(),
                styles::TEXT_BRIGHT,
            );

            if p.launch_age_s > 0.0 {
                kv(
                    &mut lines,
                    "Age:",
                    &Self::format_duration(p.launch_age_s),
                    styles::TEXT_BRIGHT,
                );
            }

            if !p.exe.is_empty() {
                let exe = if p.exe.chars().count() > 30 {
                    format!(
                        "…{}",
                        p.exe
                            .chars()
                            .rev()
                            .take(29)
                            .collect::<String>()
                            .chars()
                            .rev()
                            .collect::<String>()
                    )
                } else {
                    p.exe.to_string()
                };
                kv(&mut lines, "Exe:", &exe, styles::TEXT_BRIGHT);
            }

            if !p.role_type.is_empty() {
                kv(
                    &mut lines,
                    "Role:",
                    p.role_type.as_str(),
                    styles::ACCENT_VIOLET,
                );
            }

            if p.is_system {
                kv(&mut lines, "Type:", "System", Color::Rgb(255, 138, 76));
            }

            if !p.energy_impact.is_empty() {
                let color = match p.energy_impact.as_str() {
                    "HIGH" => styles::ACCENT_RED,
                    "MED" => Color::Rgb(255, 138, 76),
                    _ => styles::ACCENT_GREEN,
                };
                kv(&mut lines, "Energy:", p.energy_impact.as_str(), color);
            }

            let growth = self.metrics.memory_growth_rate(p.pid, 60.0);
            if growth.abs() > 1.0 {
                let color = if growth > 10.0 {
                    styles::ACCENT_RED
                } else {
                    styles::ACCENT_AMBER
                };
                kv(
                    &mut lines,
                    "Δ Mem:",
                    &format!("{:+.0}MB/min", growth),
                    color,
                );
            }
        }
        // === Empty / Nothing Selected ===
        else {
            lines.push(Line::from(vec![
                Span::raw("  "),
                Span::styled(
                    "Select a process to inspect",
                    Style::default().fg(styles::TEXT_DIM),
                ),
            ]));
        }

        // --- Actions Section ---
        lines.push(Line::from(""));
        section_header(&mut lines, "─── Actions ───────────");
        lines.push(actions_line("k", "Kill", "K", "Force Kill"));
        lines.push(actions_line("z", "Suspend", "/", "Filter"));

        let widget = Paragraph::new(lines).block(block);
        widget.render(area, buf);
    }
}

fn kv(lines: &mut Vec<Line<'static>>, key: &str, val: &str, color: Color) {
    lines.push(Line::from(vec![
        Span::raw("  "),
        Span::styled(format!("{:<8}", key), Style::default().fg(styles::TEXT_DIM)),
        Span::styled(val.to_string(), Style::default().fg(color)),
    ]));
}

fn section_header(lines: &mut Vec<Line<'static>>, text: &str) {
    lines.push(Line::from(vec![
        Span::raw("  "),
        Span::styled(text.to_string(), Style::default().fg(styles::TEXT_DIM)),
    ]));
}

fn actions_line(a1: &str, l1: &str, a2: &str, l2: &str) -> Line<'static> {
    Line::from(vec![
        Span::raw("  "),
        Span::styled(a1.to_string(), Style::default().fg(styles::ACCENT_BLUE)),
        Span::raw(format!(" {}  ", l1)),
        Span::styled(a2.to_string(), Style::default().fg(styles::ACCENT_BLUE)),
        Span::raw(format!(" {}", l2)),
    ])
}

fn titlecase(s: &str) -> String {
    let mut chars = s.chars();
    match chars.next() {
        Some(c) => c.to_uppercase().collect::<String>() + chars.as_str(),
        None => String::new(),
    }
}
